cmd/kubelet/app/server.go

```diff
@@ -412,13 +430,218 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+// On failure, the key and the certificate file will be cleaned up.
+// If the certfile or keyfile is empty, then it will use the default path, respectively:
+// /var/run/kubernetes/kubelet-client.srt, /var/run/kubernetes/kubelet-client.key.
+func getCertFromAPIServer(s *options.KubeletServer, certfile, keyfile string) (certPath, keyPath string, err error) {
+	certPath, keyPath = certfile, keyfile
+
+	// Cleanup keyfile and certfile.
+	defer func() {
+		if err != nil {
+			// Clean up cert and key.
+			if rmErr := os.Remove(certPath); rmErr != nil {
+				glog.Warningf("Failed to clean up TLS cert file %q: %v", certPath, rmErr)
+			}
+			if rmErr := os.Remove(keyPath); rmErr != nil {
+				glog.Warningf("Failed to clean up TLS private key file %q: %v", keyPath, rmErr)
+			}
+		}
+	}()
+
+	if certPath == "" {
+		certPath = defaultKubeletClientCertificateFile
+		if err := os.MkdirAll(filepath.Dir(certPath), 0644); err != nil {
+			return "", "", fmt.Errorf("unable to create dir for certificate file: %v", err)
+		}
+	}
+
+	if keyPath == "" {
+		keyPath = defaultKubeletClientKeyFile
+		if err := os.MkdirAll(filepath.Dir(keyPath), 0644); err != nil {
+			return "", "", fmt.Errorf("unable to create dir for key file: %v", err)
+		}
+	}
+
+	// (1).
+	clientConfig, err := kubeconfigClientConfig(s.BootstrapKubeconfig, s.APIServerList)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create client config: %v", err)
+	}
+	client, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create certificates signing request client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
+	if nodeIP != nil {
+		ips = []net.IP{nodeIP}
+	}
+
+	req, err := utilcertificates.NewCertificateRequest(keyPath, &pkix.Name{CommonName: hostname}, []string{hostname}, ips)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to generate certificate request: %v", err)
+	}
+
+	// (3).
+	certificate, err := requestCertificate(client, req, 3600) // Make a default timeout = 3600s
+	if err != nil {
+		return "", "", fmt.Errorf("unable to request certificate from API server: %v", err)
+	}
+
+	// (4).
+	if err = crypto.WriteCertToPath(certPath, certificate); err != nil {
+		return "", "", fmt.Errorf("unable to write certificate file: %v", err)
+	}
+
+	return certPath, keyPath, nil
+}
+
+// requestCertificate sends the certificate signing request to API server and watches the object's status.
+// It returns the API server's issued certificate (pem-encoded) on success.
+// If there is any errors, or the watch timeouts, it returns an error.
+func requestCertificate(client unversionedcertificates.CertificateSigningRequestsGetter, request []byte, defaultTimeoutSeconds int64) (certificate []byte, err error) {
+	if _, err = client.CertificateSigningRequests().Create(&certificates.CertificateSigningRequest{
```

Comment:
I am not sure how the server is supposed to verify this CSR is valid?

One option on AWS is to cross check the private key against the console output, so we could log the fingerprint to /dev/console here.  But then we'd need a way to match this CSR to an instance - is the intention to use the IP address to do that?